packages/candid/src/idl.test.ts

```diff
@@ -928,3 +933,191 @@ describe('IDL opt edge cases (embedded)', () => {
+    testSubFail(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Nat, b: IDL.Nat }));
+
+    // Opt, Null, or Reserved fields are allowed to be missing
+    testSub(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Nat, b: IDL.Opt(IDL.Nat) }));
+    testSub(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Nat, b: IDL.Null }));
+    testSub(IDL.Record({ a: IDL.Nat }), IDL.Record({ a: IDL.Nat, b: IDL.Reserved }));
+  });
+
+  describe('Subtyping on Functions', () => {
+    testReflexive(IDL.Func([], []));
+    testReflexive(IDL.Func([IDL.Nat], [IDL.Int]));
+
+    // Arg types are contravariant
+    testSub(IDL.Func([IDL.Int], []), IDL.Func([IDL.Nat], []));
+    testSub(IDL.Func([IDL.Int, IDL.Opt(IDL.Nat)], []), IDL.Func([IDL.Nat], []));
+    testSubFail(IDL.Func([IDL.Nat, IDL.Nat], []), IDL.Func([IDL.Nat], []));
+
+    // Return types are covariant
+    testSub(IDL.Func([], [IDL.Nat]), IDL.Func([], [IDL.Int]));
+    testSub(IDL.Func([], [IDL.Nat]), IDL.Func([], [IDL.Int, IDL.Opt(IDL.Nat)]));
+    testSubFail(IDL.Func([], [IDL.Nat]), IDL.Func([], [IDL.Int, IDL.Int]));
+  });
+
+  describe('Subtyping on variants', () => {
+    testReflexive(IDL.Variant({}));
+    testReflexive(IDL.Variant({ a: IDL.Nat }));
+
+    // Subtyping on individual alternatives happens pointwise
+    testSub(IDL.Variant({ a: IDL.Nat }), IDL.Variant({ a: IDL.Int }));
+    testSubFail(IDL.Variant({ a: IDL.Int }), IDL.Variant({ a: IDL.Nat }));
+
+    // Width subtyping
+    testSub(IDL.Variant({ a: IDL.Nat }), IDL.Variant({ a: IDL.Nat, b: IDL.Nat }));
+    testSubFail(IDL.Variant({ a: IDL.Nat, b: IDL.Nat }), IDL.Variant({ a: IDL.Nat }));
+  });
+
+  describe('Subtyping on services', () => {
+    testReflexive(IDL.Service({}));
+    testReflexive(IDL.Service({ f: IDL.Func([], []) }));
+
+    // Subtyping on service methods happens pointwise
+    testSub(
+      IDL.Service({ f: IDL.Func([IDL.Int], [IDL.Nat]) }),
+      IDL.Service({ f: IDL.Func([IDL.Nat], [IDL.Int]) }),
+    );
+    testSubFail(
+      IDL.Service({ f: IDL.Func([IDL.Nat], [IDL.Int]) }),
+      IDL.Service({ f: IDL.Func([IDL.Int], [IDL.Nat]) }),
+    );
+
+    // Width subtyping
+    testSub(
+      IDL.Service({ f: IDL.Func([], []), g: IDL.Func([], []) }),
+      IDL.Service({ f: IDL.Func([], []) }),
+    );
+    testSubFail(
+      IDL.Service({ f: IDL.Func([], []) }),
+      IDL.Service({ f: IDL.Func([], []), g: IDL.Func([], []) }),
+    );
+  });
+
+  describe('Subtyping on recursive types', () => {
+    const IntList = IDL.Rec();
+    IntList.fill(IDL.Opt(IDL.Record({ head: IDL.Int, tail: IntList })));
+    const NatList = IDL.Rec();
+    NatList.fill(IDL.Opt(IDL.Record({ head: IDL.Nat, tail: NatList })));
+    testSub(NatList, IntList);
+
+    const Even = IDL.Rec();
+    const Odd = IDL.Rec();
+    Even.fill(IDL.Tuple(Odd));
+    Odd.fill(IDL.Tuple(Even));
+
+    testSub(IDL.Tuple(Even), Odd);
+    testSub(IDL.Tuple(IDL.Tuple(Odd)), Odd);
+  });
+
+  describe('decoding function/service references', () => {
```

Comment:
Probably worth adding a test that a failing subtype test under an expected optional function/service type defaults to `null`.